### DISCUSSION TITLE
How do I get Renovate to fetch Terraform modules from my private registry?

### DISCUSSION THREAD

**Discussion by JvandeKraats** [renovatebot/renovate#24516] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.44.1

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

For Renovate to fetch the latest versions available of the Terraform modules that are referenced.
These modules come from our private Infrastructure Registry that is also in the GitLab self-hosted instance.

### What you tried so far.

I tried changing the `renovate.json` file to have the following content:
```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
        ":dependencyDashboard",
        ":semanticPrefixFixDepsChoreOthers",
        "group:monorepos",
        "group:recommended",
        "replacements:all",
        "workarounds:all"
    ],
    "hostRules": [
        {
            "matchHost": "aaa.bbb.cc/azure-xyz/"
        }
    ]
}

```

Where I tried to provide the token using an environment variable in my pipeline, like:  
````yaml
variables:
  RENOVATE_AAA_BBB_CC_AZURE_XYZ_TOKEN: $CI_JOB_TOKEN
````

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Datasource unauthorized (repository=group1/repo-name-1)
       "datasource": "terraform-module",
       "packageName": "aaa.bbb.cc/azure-xyz/terraform-azurerm-virtual-network/azure",
       "url": "https://aaa.bbb.cc/api/v4/packages/terraform/modules/v1/azure-xyz/terraform-azurerm-virtual-network/azure/versions"
```

</details>

I have anonymized some information in this example, that's why there is a lot of 'aaa' etc.  
Let me know if more Debug logging is needed, it takes some time to anonymize the logs, that's why I didn't just dump all my logs.  
Renovate is able to detect newer pipeline versions (that we also have defined in our own GitLab self-hosted instance).  
So it's able to see that, but as soon as it starts fetching from our private Infrastructure Registry, it's having trouble with the current configuration.

**Comment by rarkins** [comment]
Assuming that you hope for Renovate to recognise your `RENOVATE_AAA_BBB_CC_AZURE_XYZ_TOKEN` then you need to look at and/or copy/paste here the logs from when Renovate starts, because that's when it reads in such tokens if you have configured things correctly,

The `hostRules` you added is pointless and should be removed as it achieves nothing at all

**Reply by JvandeKraats** [reply]
Sure, here are the first 49 lines of my logs:
```
DEBUG: Using RE2 as regex engine
DEBUG: Parsing configs
DEBUG: Checking for config file in D:\Code/config.js
DEBUG: No config file found on disk - skipping
DEBUG: File config
       "config": {}
DEBUG: CLI config
       "config": {
         "onboarding": true,
         "platform": "gitlab",
         "endpoint": "https://aaa.bbb.cc/api/v4/",
         "autodiscover": true,
         "autodiscoverFilter": ["azure-xyz/repo-1"]
       }
DEBUG: Env config
       "config": {"hostRules": [], "token": "***********"}
DEBUG: Combined config
       "config": {
         "hostRules": [],
         "token": "***********",
         "onboarding": true,
         "platform": "gitlab",
         "endpoint": "https://aaa.bbb.cc/api/v4/",
         "autodiscover": true,
         "autodiscoverFilter": ["azure-xyz/repo-1"]
       }
DEBUG: Adding trailing slash to endpoint
DEBUG: Found valid git version: 2.41.0
DEBUG: GitLab version is: 15.6.8-ee
DEBUG: Using platform gitAuthor: Jeroen van de Kraats <jeroen.vandekraats@hotmail.com>
DEBUG: Adding token authentication for aaa.bbb.cc to hostRules
DEBUG: Using baseDir: C:/Users/j_v_d/AppData/Local/Temp/renovate
DEBUG: Using cacheDir: C:/Users/j_v_d/AppData/Local/Temp/renovate/cache
DEBUG: Using containerbaseDir: C:/Users/j_v_d/AppData/Local/Temp/renovate/cache/containerbase
DEBUG: Initializing Renovate internal cache into C:/Users/j_v_d/AppData/Local/Temp/renovate/cache/renovate/renovate-cache-v1
DEBUG: Commits limit = null
DEBUG: Setting global hostRules
DEBUG: Adding token authentication for aaa.bbb.cc to hostRules
DEBUG: validatePresets()
DEBUG: Autodiscovering GitLab repositories
DEBUG: Discovered 51 project(s)
 INFO: Autodiscovered repositories
       "length": 1,
       "repositories": ["azure-xyz/repo-1"]
DEBUG: Reinitializing hostRules for repo
DEBUG: Clearing hostRules
DEBUG: Adding token authentication for aaa.bbb.cc to hostRules
 INFO: Repository started (repository=azure-xyz/repo-1)
       "renovateVersion": "36.97.1"
```

Thanks for your quick response.

**Reply by rarkins** [reply, marked as answer]
Based on the log you can see that `RENOVATE_AAA_BBB_CC_AZURE_XYZ_TOKEN` wasn't detected/converted to hostRules. According to which feature/documentation were you hoping that it would do something? If it's https://docs.renovatebot.com/self-hosted-configuration/#detecthostrulesfromenv then I don't see you having configured that

**Reply by JvandeKraats** [reply]
Thanks for pointing me to the right part of the documentation, I wasn't able to find this specific flag myself.  
I came up with the previously mentioned 'solution' when I red [this](https://docs.renovatebot.com/configuration-options/#matchhost) part of the documentation and red somewhere you would be able to pass the token as an environment variable.

I managed to get it working by setting the environment variable `TERRAFORM__MODULE_AAA_BBB_CC_TOKEN=[Your token]`  and passing the flag `--detect-host-rules-from-env` to the CLI.
